pallets/emergency-para-xcm/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+use frame_support::traits::{ProcessMessage, QueuePausedQuery};
+use frame_system::pallet_prelude::*;
+use frame_system::RawOrigin;
+use parity_scale_codec::{Decode, Encode};
+use polkadot_parachain_primitives::primitives::{Id, RelayChainBlockNumber, XcmpMessageHandler};
+
+#[derive(Decode, Default, Encode, PartialEq, TypeInfo)]
+/// XCM Execution mode
+pub enum XcmMode {
+	/// Normal operation
+	#[default]
+	Normal,
+	/// Paused: no XCM messages are processed and `FastAuthorizedUpgrade`
+	/// origin can authorized a runtime upgrade
+	Paused,
+}
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config
+		+ cumulus_pallet_parachain_system::Config<
+			CheckAssociatedRelayNumber = Pallet<Self>,
+			XcmpMessageHandler = Pallet<Self>,
+		> + pallet_message_queue::Config<QueuePausedQuery = Pallet<Self>>
+	{
+		/// Overarching event type
+		type RuntimeEvent: From<Event> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Used before check associated relay block number. It should be what would be passed to
+		/// `cumulus_pallet_parachain_system` if this pallet was not being used.
+		type CheckAssociatedRelayNumber: CheckAssociatedRelayNumber;
+
+		/// Used to check wether message queue is paused when `XcmMode` is `Normal`. It should be
+		/// what would be passed to `pallet_message_queue` if this pallet was not being used.
+		type QueuePausedQuery: QueuePausedQuery<<Self::MessageProcessor as ProcessMessage>::Origin>;
+
+		/// The XCMP handler to be used in normal operating mode
+		type XcmpMessageHandler: XcmpMessageHandler;
+
+		/// Maximum number of relay block to skip before trigering the Paused mode.
+		type PausedThreshold: Get<RelayChainBlockNumber>;
+
+		/// Origin allowed to perform a fast authorize upgrade when XcmMode is not normal
+		type FastAuthorizeUpgradeOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// origin allowed to resume to normal operations from paused mode
+		type PausedToNormalOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event {
+		/// The XCM incoming execution was Paused
+		EnteredPausedXcmMode,
+		/// The XCM incoming execution returned to normal operation
+		NormalXcmOperationResumed,
+	}
+
+	/// An error that can occur while executing this pallet's extrinsics.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The current XCM Mode is not Paused
+		NotInPausedMode,
+	}
+
+	#[pallet::storage]
+	/// Whether incoming XCM is enabled or paused
+	pub type Mode<T: Config> = StorageValue<_, XcmMode, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			// Account for 1 read and 1 write to `Mode`
+			T::DbWeight::get().reads_writes(1, 1)
```

Comment:
This does not seem to be necessary

Reply:
It is necessary to account for `check_associated_relay_number` stuff, @elfedy can you add a comment?